apps/core/src/components/providers/ThemeProvider.tsx

```diff
@@ -12,40 +12,73 @@ interface ThemeProviderProps {
 export function ThemeProvider({ children, appId }: PropsWithChildren<ThemeProviderProps>) {
     const storageKey = `theme_${appId}`;
 
-    const getSystemTheme = () =>
-        window.matchMedia('(prefers-color-scheme: dark)').matches ? Theme.Dark : Theme.Light;
+    const getSystemTheme = () => {
+        return window.matchMedia('(prefers-color-scheme: dark)').matches ? Theme.Dark : Theme.Light;
+    };
 
-    const getInitialTheme = () => {
-        if (typeof window === 'undefined') {
-            return Theme.System;
-        } else {
-            const storedTheme = localStorage?.getItem(storageKey);
-            return storedTheme ? (storedTheme as Theme) : Theme.System;
-        }
+    const getThemePreference = () => {
+        const storedTheme = localStorage?.getItem(storageKey) as ThemePreference | null;
+        return storedTheme ? storedTheme : ThemePreference.System;
     };
 
-    const [theme, setTheme] = useState<Theme>(getInitialTheme);
+    const [systemTheme, setSystemTheme] = useState<Theme>(Theme.Light);
+    const [themePreference, setThemePreference] = useState<ThemePreference>(ThemePreference.System);
 
-    const applyTheme = useCallback((currentTheme: Theme) => {
-        const selectedTheme = currentTheme === Theme.System ? getSystemTheme() : currentTheme;
-        const documentElement = document.documentElement.classList;
-        documentElement.toggle(Theme.Dark, selectedTheme === Theme.Dark);
-        documentElement.toggle(Theme.Light, selectedTheme === Theme.Light);
-    }, []);
+    let isLoadingPreference = false;
 
+    // Load the theme values on client
     useEffect(() => {
         if (typeof window === 'undefined') return;
 
-        localStorage.setItem(storageKey, theme);
-        applyTheme(theme);
+        setSystemTheme(getSystemTheme());
+        setThemePreference(getThemePreference());
+
+        // Make the theme preference listener wait
+        // until the preference is loaded in the next render
+        isLoadingPreference = true;
+    }, []);
+
+    // When the theme preference changes..
+    useEffect(() => {
+        if (typeof window === 'undefined' || isLoadingPreference) return;
+
+        // Update localStorage with the new preference
+        localStorage.setItem(storageKey, themePreference);
+
+        // In case of SystemPreference, listen for system theme changes
+        if (themePreference === ThemePreference.System) {
+            const handleSystemThemeChange = () => {
+                const systemTheme = getSystemTheme();
+                setSystemTheme(systemTheme);
+            };
+            const systemThemeMatcher = window.matchMedia('(prefers-color-scheme: dark)');
+            systemThemeMatcher.addEventListener('change', handleSystemThemeChange);
+            return () => systemThemeMatcher.removeEventListener('change', handleSystemThemeChange);
```

Comment:
This is not a watcher, is it? it doesn't update the system theme as i change it in my config

Reply:
Fixed 🙏🏻